[TECH-482] Co2 Emissions Service

jessicaqgomez opened the pull request:
**Description:**
**1.** Four new endpoints was added to get the information about network emissions provided by immutable-insights:
Daily emissions, network emissions, network emissions per transaction and country emissions.

      Network Emissions and Network Emissions per transaction have a filter to limit the results by day, month, or year 
      depending on the range of dates in the startTime and endTime in request parameter:

       If the date range is less than or equal to one month, the results will be displayed per day, if it is greater than one month 
       but less than or equal to one year, they will be displayed per month and if it is greater than one year, they will be 
       displayed per year
       
      **e.g. Per Day**
      {
          "name": "Network Emissions Camino",
          "Filter": "Per Day",
          "value": [
            {
              "chain": "camino-columbus",
              "time": "2022-12-01",
              "value": 14991.42
            },
            {
              "chain": "camino-columbus",
              "time": "2022-12-02",
              "value": 14991.42
            },
            {
              "chain": "camino-columbus",
              "time": "2022-12-03",
              "value": 14991.42
            }
          ]
      }

     **e.g. Per month**
      {
        "name": "Network Emissions Camino",
        "Filter": "Per Month",
        "value": [
          {
            "chain": "camino-columbus",
            "time": "2022-11-01",
            "value": 13141.33
          },
          {
            "chain": "camino-columbus",
            "time": "2022-12-01",
            "value": 14991.42
          }
        ]
      }

    **e.g. Per Year**
      {
        "name": "Network Emissions Camino",
        "Filter": "Per Year",
        "value": [
          {
            "chain": "camino-columbus",
            "time": "2022-01-01",
            "value": 15168.96
          },
          {
            "chain": "camino-columbus",
            "time": "2023-01-01",
            "value": 13679.81
          }
        ]
      }
**2.** For the configuration of this endpoint, a new option was added in the services section of the config file, in which the url of the inmutableInsights service.
...
"services": {
"db": {
"dsn": "",
"driver": ""
},
"inmutableInsights": {
"urlEndpoint":"https://co2.api.immutableinsight.io",
"authorizationToken": ""
}
}
**Additionally an environment variable must be set in commandline with the auth token of inmutable insights api**
export authorizationTokenInmutable = "TOKEN"

jessicaqgomez committed:
TECH-482: co2 emissions endpoint with results filter per date

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+func stringToDate(date string) (time.Time, error) {
+	// parse a string "YYYY-MM-DD" e.g. "2022-12-25" to time.Time using timeLayout "2006-01-02" predefined in
+	// go Time package
+	t, err := time.Parse("2006-01-02", date)
```
```suggestion
	t, err := time.Parse(time.Layout, date)
```

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+	t, err := time.Parse("2006-01-02", date)
+	return t, err
+}
+func GetNetworkName(rpc string) (string, error) {
```
I would suggest to use the already implemented client in `caminogo` (api/info/client.go) in order to get the network name instead of creating an API call. You can see how magellan uses the `nodeIndexer` in the `producerChainContainer` to call the `GetContainerRange` call as an example.

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+
+const (
+	// camino chain id used in co2-api-inmutableinsights
+	columbusChainID string = "caminocolumbus"
```
Shouldn't it be dynamic depending on the `networkName` or the `networkID` that the magellan is connected to?

charalarg replied:
Maybe we need a different name for that variable, too. `ChainID` is a little bit confusing for this string.

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+	return emissions, nil
+}
+
+func CarbonIntensityFactor(chain string, startDate string, endDate string, config cfg.EndpointService) ([]*models.EmissionsResult, error) {
```
I would make the function private

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+	}
+	return parsedValue
+}
+func GetAvgEmissionsFilter(e []*models.EmissionsResult, perYear bool) ([]*models.EmissionsResult, error) {
```
I would make the function private

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+
+	return response, nil
+}
+func GetAvgEmissionsValue(e []*models.EmissionsResult) float64 {
```
I would make the function private

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+
+	return response, nil
+}
+func AccesibleChains(config cfg.EndpointService) ([]string, error) {
```
I would make the function private

charalarg replied:
There is also a typo in `AccesibleChains` :)

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+	return response, nil
+}
+
+func Country(country string, startDate string, endDate string, config cfg.EndpointService) ([]*models.EmissionsResult, error) {
```
I would make the function private

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+	if err != nil {
+		return response, err
+	}
+	if config.AuthorizationToken != "" {
```
I would remove this check and let the function return the error response back as it would be if the token was wrong and not empty. Same for the same check in the four helper functions bellow.

jessicaqgomez replied:
ok, it's solved

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+			})
+		}
+	}
+	if dailyEmissions == nil {
```
This check could be ommited

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+		err           error
+	)
+	for _, country := range countryIDs {
+		countryInfo, errCountry := Country(country, startDatef, endDatef, config)
```
I would rename `errCountry` with `err` and  remove both the `var err error` statement above and the else clause bellow.

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,429 @@
+			err = errCountry
+		}
+	}
+	if emissionsInfo == nil {
```
This could be omitted and return the `err` variable in return bellow instead of nil.

charalarg reviewed (changes requested):
In general it would be perfect if we have some working and failing tests for the emissions functionality, in order to test these calculations covering the if-else/switch clauses and also be sure that everything is working fine along with the external connections to the third-party endpoints.

jessicaqgomez commented:
@charalarg I fixed all the recommendations and corrections in the last commit, please take a look when you have time

peak3d committed:
Update dependencies
- New genesis (depositOffer memo)
- Removal of AddressStateRegisterNode
- Make multisig addresses varchar (b58 has var length)

charalarg commented on utils/emissions.go:
```diff
@@ -0,0 +1,368 @@
+
+func GetCountryEmissions(startDate time.Time, endDate time.Time, config cfg.EndpointService, rpc string) (models.Emissions, error) {
+	var (
+		err         error
```
This could be omitted i think. And replace `countryInfo, err = country(countryID, startDatef, endDatef, config)` with `countryInfo, err := country(countryID, startDatef, endDatef, config)`
